jetty-ee10/jetty-ee10-webapp/src/main/java/org/eclipse/jetty/ee10/webapp/MetaInfConfiguration.java

```diff
@@ -378,64 +378,70 @@ public void scanForResources(WebAppContext context, Resource target, ConcurrentH
         // Resource target does not exist
         if (Resources.missing(target))
             return;
-        ResourceFactory resourceFactory = ResourceFactory.of(context);
 
-        Resource resourcesDir;
-        if (cache != null && cache.containsKey(target))
-        {
-            resourcesDir = cache.get(target);
-            if (isEmptyResource(resourcesDir))
-            {
-                if (LOG.isDebugEnabled())
-                    LOG.debug("{} cached as containing no META-INF/resources", target);
-                return;
-            }
-            else if (LOG.isDebugEnabled())
-                LOG.debug("{} META-INF/resources found in cache ", target);
-        }
-        else
-        {
-            //not using caches or not in the cache so check for the resources dir
-            if (LOG.isDebugEnabled())
-                LOG.debug("{} META-INF/resources checked", target);
-            if (target.isDirectory())
-            {
-                //TODO think  how to handle an unpacked jar file (eg for osgi)
-                resourcesDir = target.resolve("/META-INF/resources");
-            }
-            else
-            {
-                // Resource represents a packed jar
-                URI uri = target.getURI();
-                resourcesDir = resourceFactory.newResource(URIUtil.uriJarPrefix(uri, "!/META-INF/resources"));
-            }
-
-            if (Resources.isReadableDirectory(resourcesDir) && (cache != null))
-            {
-                Resource old = cache.putIfAbsent(target, resourcesDir);
-                if (old != null)
-                    resourcesDir = old;
-                else if (LOG.isDebugEnabled())
-                    LOG.debug("{} META-INF/resources cache updated", target);
-            }
-
-            if (isEmptyResource(resourcesDir))
-            {
-                return;
-            }
-        }
-
-        //add it to the meta inf resources for this context
-        Set<Resource> dirs = (Set<Resource>)context.getAttribute(METAINF_RESOURCES);
-        if (dirs == null)
-        {
-            dirs = new HashSet<>();
-            context.setAttribute(METAINF_RESOURCES, dirs);
-        }
-        if (LOG.isDebugEnabled())
-            LOG.debug("{} added to context", resourcesDir);
-
-        dirs.add(resourcesDir);
+       try (ResourceFactory.Closeable closeable = ResourceFactory.closeable())
```

Comment:
This needs to be moved out of the individual scan methods and done just once outside of the outermost loop.   Either pass in the resource factory or resolve the non directory resources to mounted resources before calling the method.

Reply:
Agreed that there should be one `ResourceFactory.Closeable` for the `scanJars` method to convert everything that's a `.jar` into a `file:jar:` `Resource`. This change is forthcoming.